### PR TITLE
Grids inherit from Graphs

### PR DESCRIPTION
**Pull request by mcflugen**
This pull request refactors landlab *ModelGrids* (*RasterModelGrid*, *HexModelGrid*, etc.) to inherit functionality from corresponding *Graph* classes. There is a lot here but, for the most part, it is simply removing all of the topology-related code (e.g. *nodes_at_link*, *patches_at_node*, etc.) from the grid classes and using the methods from the graph classes.

I've also removed many functions/methods that were labeled as deprecated.

Most of the tests are passing (at least for me locally) but there are still a few remaining tests that are failing that @kbarnhart and @gregtucker could probably help out with. I believe the failures are related to the updated way that rectangular hex grids are represented. In landlab-1.0, a hex grid looked like,
```
  * - * - * - * - *
 / \ / \ / \ / \ /|
* - * - * - * - * |
|\ / \ / \ / \ / \|
| * - * - * - * - *
|/ \ / \ / \ / \ /
* - * - * - * - *
```
Now in landlab-2.0 they look like (notice the two missing vertical links),
```
  * - * - * - * - *
 / \ / \ / \ / \ /
* - * - * - * - *
 \ / \ / \ / \ / \
  * - * - * - * - *
 / \ / \ / \ / \ /
* - * - * - * - *
```
I realize this pull request is far too large and I apologize for that, I got carried away.

### CODE REVIEW

**Comment by kbarnhart** [comment]
I just took a look at the failing tests for this and found that the failing tests in components, utils, and ca are in the following files (additional failing tests are in the graph and grid submodule):

I see: 48 failed, 2761 passed, 13 skipped, 122 warnings, 4 error 

```
landlab/ca/boundaries/hex_lattice_tectonicizer.py ....FFFFF..F..         [  1%]
landlab/ca/boundaries/tests/test_hex_normal_fault.py FF                  [  1%]
landlab/ca/tests/grain_hill.py F.                                        [  1%]
landlab/components/discharge_diffuser/diffuse_by_discharge.py F.         [  2%]
landlab/components/flow_routing/tests/test_flow_routing.py .........FF   [ 11%]
landlab/components/lake_fill/lake_fill_barnes.py .............F..        [ 12%]
landlab/components/potentiality_flowrouting/route_flow_by_boundary.py F  [ 17%]
landlab/components/potentiality_flowrouting/route_flow_by_boundary_1st.py F [ 17%]
landlab/components/profiler/tests/test_channel_profile.py FFFFFFEEEEFFFF [ 18%]
FFF
landlab/utils/tests/test_distance_from_divide.py .FFFFFF                 [ 92%]
landlab/utils/distance_to_divide.py F                                    [ 89%]
```

Starting to tackle who should work on what. 
- [ ] ca module: @gregtucker 
- [ ] Diffuse by discharge: @SiccarPoint 
- [ ] Test flow routing: me
- [ ] Lake fill barnes: @SiccarPoint 
- [ ] Channel Profiler: me
- [ ] Potentiality flow routing: @SiccarPoint 
- [ ] Distance to divide: me
- [ ] Errors in graph and grid: @mcflugen 

I might have missed a few places where there are errors, but this is at least a start.

**Comment by mcflugen** [comment]
Only six failures left - just about there. Thank you @gregtucker and @kbarnhart for your help! The only remaining tests that need fixing are in the following:
* [ ] `landlab/components/discharge_diffuser/diffuse_by_discharge.py`
* [ ] `landlab/components/flow_routing/tests/test_flow_routing.py`
* [ ] `landlab/components/lake_fill/lake_fill_barnes.py`
* [ ] `landlab/components/potentiality_flowrouting/route_flow_by_boundary.py`
* [ ] `landlab/components/potentiality_flowrouting/route_flow_by_boundary_1st.py`

@SiccarPoint it appears these are tests you might be familiar with. If you could have a look at these it would be much appreciated!

**Comment by kbarnhart** [comment]
@mcflugen I'm taking a stab at some of the tests right now. A few things I've done:

1) Removed a file with a failing test that will get removed in #956 (see #957).
2) Found a few places in the grid documentation where duplicate/incorrect information in input parameters was located (and fixed).
3) Tried to figure out what's up with the potentiality flow routing error of @SiccarPoint 

First, with graph to grid branch:
```python
mg = HexModelGrid((4, 6), 
    spacing=2., 
    node_layout="rect", 
    orientation="vertical")

mg.x_of_node
array([ 0.        ,  3.46410162,  6.92820323,  1.73205081,  5.19615242,
        8.66025404,  0.        ,  3.46410162,  6.92820323,  1.73205081,
        5.19615242,  8.66025404,  0.        ,  3.46410162,  6.92820323,
        1.73205081,  5.19615242,  8.66025404,  0.        ,  3.46410162,
        6.92820323,  1.73205081,  5.19615242,  8.66025404])

mg.y_of_node  
array([ 0.,  0.,  0.,  1.,  1.,  1.,  2.,  2.,  2.,  3.,  3.,  3.,  4.,
        4.,  4.,  5.,  5.,  5.,  6.,  6.,  6.,  7.,  7.,  7.])

mg.status_at_node
array([1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1], dtype=uint8)
       
mg.status_at_link
array([4, 4, 4, 4, 4, 4, 0, 0, 4, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 4, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 4, 0,
       0, 4, 4, 4, 4, 4, 4], dtype=uint8)
  
mg.adjacent_nodes_at_node
array([[ 3,  6, -1, -1, -1, -1],
       [ 4,  7,  3, -1, -1, -1],
       [ 5,  8,  4, -1, -1, -1],
       [ 7,  9,  6,  0,  1, -1],
       [ 8, 10,  7,  1,  2, -1],
       [11,  8,  2, -1, -1, -1],
       [ 9, 12,  0,  3, -1, -1],
       [10, 13,  9,  3,  1,  4],
       [11, 14, 10,  4,  2,  5],
       [13, 15, 12,  6,  3,  7],
       [14, 16, 13,  7,  4,  8],
       [17, 14,  8,  5, -1, -1],
       [15, 18,  6,  9, -1, -1],
       [16, 19, 15,  9,  7, 10],
       [17, 20, 16, 10,  8, 11],
       [19, 21, 18, 12,  9, 13],
       [20, 22, 19, 13, 10, 14],
       [23, 20, 14, 11, -1, -1],
       [21, 12, 15, -1, -1, -1],
       [22, 21, 15, 13, 16, -1],
       [23, 22, 16, 14, 17, -1],
       [18, 15, 19, -1, -1, -1],
       [19, 16, 20, -1, -1, -1],
       [20, 17, -1, -1, -1, -1]])
       
```
Second with v2_dev branch (which passes)
```python
mg = HexModelGrid(4, 6, dx=2., node_layout='rect', orientation='vertical')

mg.x_of_node        
array([ 0.        ,  3.46410162,  6.92820323,  1.73205081,  5.19615242,
        8.66025404,  0.        ,  3.46410162,  6.92820323,  1.73205081,
        5.19615242,  8.66025404,  0.        ,  3.46410162,  6.92820323,
        1.73205081,  5.19615242,  8.66025404,  0.        ,  3.46410162,
        6.92820323,  1.73205081,  5.19615242,  8.66025404])
        
mg.y_of_node
array([ 0.,  0.,  0.,  1.,  1.,  1.,  2.,  2.,  2.,  3.,  3.,  3.,  4.,
        4.,  4.,  5.,  5.,  5.,  6.,  6.,  6.,  7.,  7.,  7.])
        
mg.status_at_node
array([1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1], dtype=uint8)
       
mg.status_at_link
array([4, 4, 4, 4, 4, 4, 4, 4, 0, 0, 4, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0,
       4, 0, 0, 4, 4, 4, 4, 4, 4, 4, 4], dtype=uint8)
       
mg.adjacent_nodes_at_node
array([[ 1,  3,  6, -1, -1, -1],
       [ 2,  4,  7,  3,  0, -1],
       [ 5,  8,  4,  1, -1, -1],
       [ 7,  9,  6,  0,  1, -1],
       [ 8, 10,  7,  1,  2, -1],
       [11,  8,  2, -1, -1, -1],
       [ 9, 12,  0,  3, -1, -1],
       [10, 13,  9,  3,  1,  4],
       [11, 14, 10,  4,  2,  5],
       [13, 15, 12,  6,  3,  7],
       [14, 16, 13,  7,  4,  8],
       [17, 14,  8,  5, -1, -1],
       [15, 18,  6,  9, -1, -1],
       [16, 19, 15,  9,  7, 10],
       [17, 20, 16, 10,  8, 11],
       [19, 21, 18, 12,  9, 13],
       [20, 22, 19, 13, 10, 14],
       [23, 20, 14, 11, -1, -1],
       [21, 12, 15, -1, -1, -1],
       [22, 21, 15, 13, 16, -1],
       [23, 22, 16, 14, 17, -1],
       [22, 18, 15, 19, -1, -1],
       [23, 21, 19, 16, 20, -1],
       [22, 20, 17, -1, -1, -1]])
```

Only thing that changes in the removal of the links (as you point out above). Since this is an open grid, I suspect those links were used... I'll wait for @SiccarPoint to comment.

**Comment by kbarnhart** [comment]
@mcflugen. I've been looking at the skipped tests too. Some of them are test for deprecated or unsupported things. I've removed those. 

Those that remain are mostly in the graph/structured_quad module. 

```python
@mark.skip("speed tests")
@mark.parametrize(
    "method",
    (
        "links_at_patch",
        "nodes_at_link",
        "horizontal_links",
        "vertical_links",
        "perimeter_nodes",
        "links_at_node",
        "patches_at_link",
        "link_dirs_at_node",
        "patches_at_node",
    ),
)
@mark.parametrize("size", (10, 11))
def test_layouts_cython_is_faster(method, size):
    from timeit import timeit

    n_rows, n_cols = 3 * 2 ** size, 4 * 2 ** size

    def time_method(impl):
        return timeit(
            "{impl}.{method}(({n_rows}, {n_cols}))".format(
                impl=impl, method=method, n_rows=n_rows, n_cols=n_cols
            ),
            setup="from landlab.graph.structured_quad.structured_quad import {impl}".format(
                impl=impl
            ),
            number=1,
        )

    benchmark = time_method("StructuredQuadLayoutPython")
    time = time_method("StructuredQuadLayoutCython")
    speedup = benchmark / time

    assert speedup > 1  # or time < 1e-2
```

I tried uncommenting this and found that 5 of them failed. Do you think this matters? Should these tests pass? Should they be removed?

**Comment by kbarnhart** [comment]
I've also gotten rid of the lint in graph and grid (with one exception)

@mcflugen I don't know how to address this one:
```
landlab/graph/object/at_patch.py:32:5: F821 undefined name '_get_links_at_patch'
```

**Comment by kbarnhart** [comment]
@mcflugen I've looked carefully at the remaining failing tests. One (in lake_mapper) was a very small decimal error (I've fixed). The one in discharge diffuser turned out to be a vestigial copy of the potentiality flow routing test (I've removed).

The only failing tests (besides the potentiality flow routing test that remains and @SiccarPoint should look at) are the two tests in `route_flow`. I'm assigning these to you since they have to do with the number and layout of RadialModelGrid nodes not actually being the same after this PR. 

Both of these tests are in `landlab/components/flow_routing/tests/test_flow_routing` and are:
- `test_voronoi` and
- `test_voronoi_closedinternal`

They are both based on a Radial Model Grid with the following original instantiation: 

 ```python
vmg = RadialModelGrid(2, dr=2.0)
```
This resulted in the following grid.
Screen Shot 2019-10-01 at 2 27 09 PM

The code in the test presently reads

```python
vmg = RadialModelGrid(n_rings=2, nodes_in_first_ring=3)
```

Which gives the following grid:
Screen Shot 2019-10-01 at 2 25 23 PM

This is clearly wrong and should have six nodes in the first ring. 

```python
vmg = RadialModelGrid(n_rings=2, nodes_in_first_ring=6)
```
Which gives 
Screen Shot 2019-10-01 at 2 31 22 PM

You can see that the old RadialModelGrid created one extra node (just above east).

I can see why the new RadialModelGrid is better. 

I can revise this test to functionally test the same properties of the code but reflect the new RadialModelgrid... But first I want to make sure that the revised change to RadialModelGrid is intentional.

**Comment by mcflugen** [comment]
@kbarnhart Thanks for the help! I've fixed the last of the tests, which was the *RadialModelGrid*. Please have a look to see if what I did makes sense. I'm now mostly dealing with Windows errors where arrays in doctests are of *dtype=np.int32* rather that just a usual *int* for that system. Anyway, I think I know how to fix it and it shouldn't take too long.

It looks like there may be an occasional segfault in one of the cython bits of code. I'm not sure what that's all about.

**Comment by kbarnhart** [comment]
@mcflugen nice work!

**Comment by kbarnhart** [comment]
================================== FAILURES ===================================
current failure is only on AppVeyor and is:
```bash
3137__________________________ test_graph_nodes_property __________________________
3138Traceback (most recent call last):
3139  File "C:\projects\landlab\tests\graph\test_graph.py", line 58, in test_graph_nodes_property
3140    graph.nodes[0] = 99
3141  File "c:\miniconda37-x64\lib\site-packages\_pytest\python_api.py", line 747, in __exit__
3142    fail(self.message)
3143  File "c:\miniconda37-x64\lib\site-packages\_pytest\outcomes.py", line 128, in fail
3144    raise Failed(msg=msg, pytrace=pytrace)
3145Failed: DID NOT RAISE <class 'ValueError'>
3146============================== warnings summary===============================
3147
```

File: landlab/io/netcdf/write.py [hunk 1]
```diff
@@ -322,11 +322,11 @@ def _add_cell_spatial_variables(root, grid, **kwds):
```
Comment (kbarnhart) [inline landlab/io/netcdf/write.py]: beyond the scope of this PR... but should we change the write netcdf to be based on xarray now?

File: landlab/io/shapefile/read_shapefile.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def read_shapefile(file, dbf=None, store_polyline_vertices=True):
     array([  0.,   5.,  10.,  10.])
     >>> grid.nodes_at_link
     array([[0, 1],
-           [1, 2],
+           [2, 1],
```
Comment (kbarnhart) [inline landlab/io/shapefile/read_shapefile.py]: Should I assume that it is OK that the order changed...

File: landlab/io/netcdf/write.py (hunk 1 above)
Comment (mcflugen) [reply landlab/io/netcdf/write.py]: Yes, definitely! I could do a separate pull request for it.

File: landlab/components/depression_finder/lake_mapper.py [hunk 3]
```diff
@@ -219,8 +219,8 @@ def __init__(self, grid, routing="D8", pits="flow__sink_flag", reroute_flow=True
 
         if isinstance(grid, RasterModelGrid) and (routing == "D8"):
             self._D8 = True
-            self._num_nbrs = 8
-            self._diag_link_length = np.sqrt(grid._dx ** 2 + grid._dy ** 2)
+            self.num_nbrs = 8
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/depression_finder/lake_mapper.py]: Should have been more specific. `self.num_nbrs` should be `self._num_nbrs`

File: landlab/components/diffusion/diffusion.py [hunk 4]
```diff
@@ -286,8 +291,8 @@ def __init__(self, grid, linear_diffusivity=0.01, method="simple", deposit=True)
         if self._use_patches or self._kd_on_links:
             mg = self._grid
             try:
-                self._hoz = self._grid.horizontal_links.flatten()
-                self._vert = self._grid.vertical_links.flatten()
+                self._hoz = self.grid.horizontal_links
+                self._vert = self.grid.vertical_links
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/diffusion/diffusion.py]: At the moment, ALL other components use `self._grid` wherever they refer to the grid internally. 

My thinking was that they inherit from `Component`, and all call `super` and thus they have the `self._grid` attribute.

File: landlab/io/netcdf/write.py (hunk 1 above)
Comment (kbarnhart) [reply landlab/io/netcdf/write.py]: I'll make an issue for this to keep track.

File: landlab/io/shapefile/read_shapefile.py (hunk 2 above)
Comment (kbarnhart) [reply landlab/io/shapefile/read_shapefile.py]: Cool. Thanks for clarifying.

File: landlab/io/netcdf/write.py (hunk 1 above)
Comment (kbarnhart) [reply landlab/io/netcdf/write.py]: See Issue #1014

File: landlab/components/overland_flow/generate_overland_flow_deAlmeida.py [hunk 5]
```diff
@@ -91,10 +91,65 @@
...
+
+    if len(args) == 0:
+        return active_links_at_node.T
+    elif len(args) == 1:
+        node_ids = np.broadcast_arrays(args[0])[0]
+        return active_links_at_node[node_ids, :].T
+    else:
+        raise ValueError("only zero or one arguments accepted")
+
+
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/overland_flow/generate_overland_flow_deAlmeida.py]: We have this already on #925.

**Comment by mcflugen** [comment]
@kbarnhart that you so much for your help with this! 👏

File: landlab/components/flow_director/flow_director_d8.py [hunk 6]
```diff
@@ -12,7 +12,7 @@
 
 import numpy
 
-from landlab import FIXED_GRADIENT_BOUNDARY, FIXED_VALUE_BOUNDARY, VoronoiDelaunayGrid
+from landlab import FIXED_GRADIENT_BOUNDARY, FIXED_VALUE_BOUNDARY
```
[the remark this answers is not in the document]
Comment (mcflugen) [reply landlab/components/flow_director/flow_director_d8.py]: Yes, probably so. I'll add some.

[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/flow_director/flow_director_d8.py]: cool!

File: landlab/components/diffusion/diffusion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/diffusion/diffusion.py]: `self.grid` is documented [here](https://github.com/landlab/landlab/blob/3305756b50cb4e3769bce42853dabbdab4de61d2/landlab/core/model_component.py#L445)

`self._grid` is undocumented.

File: landlab/components/flow_director/flow_director_d8.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcflugen) [reply landlab/components/flow_director/flow_director_d8.py]: @kbarnhart I've added some docstrings for the `BC_*` attributes.

Is it OK if I change all of these to use `grid.BC_` in another pull request?

[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/components/flow_director/flow_director_d8.py]: for sure, as long as we make an issue to remember before merging this PR.

**Comment by kbarnhart** [comment]
@mcflugen I'm merging this so I can bring it into the docs and start fixing those errors.

**Pull request merged**